### PR TITLE
HDDS-7116. Avoid leaking RocksObject from DBProfile

### PR DESCRIPTION
**Pull request by duongnguyen0**
## What changes were proposed in this pull request?

DBProfile and its related RocksDB settings components creates a number of RocksObjects which are left unclosed.
This PR is a result of hunting those RocksObjects and ensure they're closed before GC.

Key changes:
- In RocksDatabase, many `TableConfig` objects is created and then filtered (org.apache.hadoop.hdds.utils.db.RocksDatabase#open). Each [TableConfig](https://github.com/duongnguyen0/ozone/blob/a43b3da2ab8c9e5684436e7c89adb228712068e7/hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/TableConfig.java#L35-L35) object encloses a ColumnFamilyOptions instance. The pattern of creating before filtering needs to be inverse.
- ColumnFamilyOptions itself is RocksObject but also consists of a hierarchy of other RocksObjects which are not closed automatically by `ColumnFamilyOptions.close`. We need to explicitly traverse through the child objects and close them.
- Sometimes, ColumnFamilyOptions is created before its content (which is other RocksObject) is replaced, [e.g.](https://github.com/apache/ozone/blob/master/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/utils/db/DatanodeDBProfile.java#L124). This leaves unclosed RocksObject being GCed. 
- DBStoreBuilder create a lots of immediate TableConfig object and doesn't close them.
https://issues.apache.org/jira/browse/HDDS-7116

## How was this patch tested?

Standard CI: https://github.com/duongnguyen0/ozone/actions/runs/2836891145

### CODE REVIEW

**Comment by duongnguyen0** [comment]
@szetszwo This is a split from #3646, would be nice if you can have a look.

**Review by szetszwo — COMMENTED** [review]
@duongnguyen0 , thanks for working on this.  Some comments inlined.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedColumnFamilyOptions.java [hunk 1]
```diff
@@ -36,6 +36,16 @@ public ManagedColumnFamilyOptions(ColumnFamilyOptions columnFamilyOptions) {
   @Override
   public ManagedColumnFamilyOptions setTableFormatConfig(
```
[the remark this answers is not in the document]
Comment (duongnguyen0) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedColumnFamilyOptions.java:37]: Right.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RocksDatabase.java [hunk 2]
```diff
@@ -405,7 +400,7 @@ public ColumnFamily getColumnFamily(String key) {
     return columnFamilies.get(key);
   }
 
-  public Collection<ColumnFamily> getColumnFamilies() {
+  public Collection<ColumnFamily> getExtraColumnFamilies() {
```
[the remark this answers is not in the document]
Comment (duongnguyen0) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RocksDatabase.java:403]: Yup, I renamed the wrong methods. Thanks for noticing it.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedColumnFamilyOptions.java [hunk 3]
```diff
@@ -36,13 +36,48 @@ public ManagedColumnFamilyOptions(ColumnFamilyOptions columnFamilyOptions) {
...
```
Comment (szetszwo) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/managed/ManagedColumnFamilyOptions.java:37]: We should synchronize this method.

**Review by szetszwo — APPROVED** [review]
@duongnguyen0 , thanks for the update.  Just a minor comment inlined.  Could you address it in the next pull request?

+1 the change looks good.

**Comment by duongnguyen0** [comment]
> @duongnguyen0 , thanks for the update. Just a minor comment inlined. Could you address it in the next pull request?

Thanks for the detailed review @szetszwo . I address the comment in the new PR #3678.

**Comment by ChenSammi** [comment]
Hey @duongkame , thanks for working on the RocksDB object clean close thing.  This patch looks like changed one function of Ozone DN, sharing LRU cache between different RocksDB instances.  Here is the object list dumped from one DN. 

image

**Comment by kerneltime** [comment]
I believe the sharing of the cache was added to limit the DB memory growth. I think it is prudent to estimate the resulting memory growth from not sharing the LRU cache across DB instances. With merged RocksDB it should be an issue but customers upgrading would still need rocksDB per container.

**Comment by ChenSammi** [comment]
@duongkame , yes, we need to share an LRU cache instance between RocksDB, for both merged RocksDB and rocksDB per container. So the cache can be fully utilized and also RocksDB memory consumption can be capped.   

The AtomicReference field of previous implementation is for this purpose.  I would suggest bringing this part of implementation back.

**Comment by lokeshj1703** [comment]
@duongkame We should also add back the TestKeyValueContainer#testContainersShareColumnFamilyOptions.

**Comment by duongkame** [comment]
@ChenSammi @lokeshj1703 I created https://issues.apache.org/jira/browse/HDDS-7362, will work on that soon.

**Comment by ChenSammi** [comment]
Hey @captainzmc I would suggest to revert this commit on 1.3 branch, for it may cause a lot of DN memory used for RocksDB cache, and DN may OOM at last.

**Comment by captainzmc** [comment]
Thanks for @ChenSammi 's reminder. Yes, if it affects the stability of DN, we should revert it in 1.3 branch.

**Comment by duongkame** [comment]
@ChenSammi @captainzmc  we can't just revert this commit without reverting all the work on RocksDB managed object and RocksDB upgrade. And that would result in another RocksDb stability issue that's solved by the RocksDB upgrade.

Lemme work on [HDDS-7362](https://issues.apache.org/jira/browse/HDDS-7362) to bring back the share of LRU cache across RocksDB instances.

**Comment by duongkame** [comment]
PR to have LRU cache shared between RocksDB. https://github.com/apache/ozone/pull/3921

**Pull request merged**